src/item/enchantment/EnchantmentHelper.php

```diff
@@ -0,0 +1,248 @@
+					$spaceZ = max(min($z, 1), -1);
+
+					$lowerSpaceBlock = $world->getBlock($tablePos->add($spaceX, 0, $spaceZ));
+					if($lowerSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+					$upperSpaceBlock = $world->getBlock($tablePos->add($spaceX, 1, $spaceZ));
+					if($upperSpaceBlock->getTypeId() !== BlockTypeIds::AIR){
+						break;
+					}
+
+					$bookshelfCount++;
+					if($bookshelfCount === self::MAX_BOOKSHELF_COUNT){
+						return $bookshelfCount;
+					}
+				}
+			}
+		}
+
+		return $bookshelfCount;
+	}
+
+	private static function createEnchantOption(Random $random, Item $inputItem, int $optionCost, int $slot) : EnchantOption{
+		$cost = $optionCost;
+
+		$enchantability = self::getEnchantability($inputItem);
+		$cost = $cost + $random->nextRange(0, $enchantability >> 2) + $random->nextRange(0, $enchantability >> 2) + 1;
+		// Random bonus for enchantment cost between 0.85 and 1.15
+		$bonus = 1 + ($random->nextFloat() + $random->nextFloat() - 1) * 0.15;
+		$cost = (int) ($cost * $bonus);
+
+		$resultEnchantments = [];
+		$availableEnchantments = self::getAvailableEnchantments($cost, $inputItem);
+
+		if(count($availableEnchantments) !== 0){
+			/** @var EnchantmentInstance $lastEnchantment */
+			$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+			$resultEnchantments[] = $lastEnchantment;
+
+			// With probability (cost + 1) / 50, continue adding enchantments
+			while($random->nextFloat() <= ($cost + 1) / 50){
+				// Remove from the list of available enchantments anything that conflicts
+				// with previously-chosen enchantments
+				$availableEnchantments = array_filter(
+					$availableEnchantments,
+					fn(EnchantmentInstance $e) => $e->getType() !== $lastEnchantment->getType() &&
+						$e->getType()->isCompatibleWith($lastEnchantment->getType())
+				);
+
+				if(count($availableEnchantments) === 0){
+					break;
+				}
+
+				/** @var EnchantmentInstance $lastEnchantment */
+				$lastEnchantment = self::getRandomWeightedEnchantment($random, $availableEnchantments);
+				$resultEnchantments[] = $lastEnchantment;
+
+				$cost >>= 1;
+			}
+		}
+
+		$protocolEnchantments = array_map(
+			fn(EnchantmentInstance $e) => new Enchant(EnchantmentIdMap::getInstance()->toId($e->getType()), $e->getLevel()),
+			$resultEnchantments
+		);
+
+		return new EnchantOption($optionCost, $slot, $protocolEnchantments, [], [], self::getRandomOptionName($random), $slot);
+	}
+
+	private static function getEnchantability(Item $item) : int{
+		if($item instanceof TieredTool){
+			return match ($item->getTier()) {
+				ToolTier::WOOD() => 15,
+				ToolTier::STONE() => 5,
+				ToolTier::IRON() => 14,
+				ToolTier::GOLD() => 22,
+				ToolTier::DIAMOND() => 10,
+				ToolTier::NETHERITE() => 15,
+				default => 1
+			};
+		}
+
+		return match ($item->getTypeId()) {
+			ItemIds::LEATHER_CAP, ItemIds::LEATHER_TUNIC, ItemIds::LEATHER_PANTS, ItemIds::LEATHER_BOOTS => 15,
```

Comment:
Yeah, but unfortunately, it will be necessary to write enchantability for each armor item separately, because ArmorTypeInfo is different for each of them. In the future, we could create an ArmorMaterial (like ToolTier) where enchantability is specified for each armor material

Reply:
Or can I create an ArmorMaterial and add the optional $material to ArmorTypeInfo (default will be leather)?